File: app/src/main/java/com/example/kineticpulsemobileapp/BLEManager.java
package com.example.kineticpulsemobileapp;

import android.Manifest;
import android.bluetooth.BluetoothAdapter;
import android.bluetooth.BluetoothDevice;
import android.bluetooth.BluetoothGatt;
import android.bluetooth.BluetoothGattCallback;
import android.bluetooth.BluetoothGattCharacteristic;
import android.bluetooth.BluetoothGattDescriptor;
import android.bluetooth.BluetoothProfile;
import android.bluetooth.le.BluetoothLeScanner;
import android.bluetooth.le.ScanCallback;
import android.bluetooth.le.ScanFilter;
import android.bluetooth.le.ScanResult;
import android.bluetooth.le.ScanSettings;
import android.content.Context;
import android.content.pm.PackageManager;
import android.os.Build;
import android.os.Handler;
import android.os.Looper;
import android.os.ParcelUuid;
import android.util.Log;
import androidx.core.app.ActivityCompat;
import org.json.JSONObject;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

public class BLEManager {
    private static final String TAG = "BLEManager";

    // Nordic UART Service UUIDs
    private static final UUID NUS_SERVICE_UUID = UUID.fromString("6E400001-B5A3-F393-E0A9-E50E24DCCA9E");
    private static final UUID NUS_TX_UUID = UUID.fromString("6E400003-B5A3-F393-E0A9-E50E24DCCA9E"); // ESP → Phone (notifications)
    private static final UUID NUS_RX_UUID = UUID.fromString("6E400002-B5A3-F393-E0A9-E50E24DCCA9E"); // Phone → ESP (write)
    private static final UUID CLIENT_CHARACTERISTIC_CONFIG = UUID.fromString("00002902-0000-1000-8000-00805f9b34fb");

    private final Context context;
    private final BLECallback callback;
    private BluetoothAdapter bluetoothAdapter;
    private BluetoothLeScanner scanner;
    private BluetoothGatt bluetoothGatt;
    private BluetoothGattCharacteristic txCharacteristic;
    private BluetoothGattCharacteristic rxCharacteristic;

    private boolean isScanning = false;
    private boolean isConnected = false;
    private final Handler handler = new Handler(Looper.getMainLooper());
    private static final long SCAN_PERIOD = 10000; // 10 seconds

    public interface BLECallback {
        void onDeviceFound(BluetoothDevice device, String name);
        void onConnected();
        void onDisconnected();
        void onDataReceived(String data);
        void onError(String error);
    }

    public BLEManager(Context context, BLECallback callback) {
        this.context = context;
        this.callback = callback;
        this.bluetoothAdapter = BluetoothAdapter.getDefaultAdapter();
        if (bluetoothAdapter != null) {
            this.scanner = bluetoothAdapter.getBluetoothLeScanner();
        }
    }

    public boolean isBluetoothEnabled() {
        return bluetoothAdapter != null && bluetoothAdapter.isEnabled();
    }

    public void startScan() {
        if (!isBluetoothEnabled() || scanner == null) {
            callback.onError("Bluetooth not available or not enabled");
            return;
        }

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            if (ActivityCompat.checkSelfPermission(context, Manifest.permission.BLUETOOTH_SCAN)
                    != PackageManager.PERMISSION_GRANTED) {
                callback.onError("Bluetooth scan permission not granted");
                return;
            }
        }

        if (isScanning) {
            Log.d(TAG, "Already scanning");
            return;
        }

        // Scan filters for NUS service
        List<ScanFilter> filters = new ArrayList<>();
        ScanFilter filter = new ScanFilter.Builder()
                .setServiceUuid(new ParcelUuid(NUS_SERVICE_UUID))
                .build();
        filters.add(filter);

        ScanSettings settings = new ScanSettings.Builder()
                .setScanMode(ScanSettings.SCAN_MODE_LOW_LATENCY)
                .build();

        isScanning = true;
        scanner.startScan(filters, settings, scanCallback);
        Log.d(TAG, "🔍 Started BLE scan for NUS service");

        // Stop scan after SCAN_PERIOD
        handler.postDelayed(() -> {
            if (isScanning) {
                stopScan();
                callback.onError("No devices found");
            }
        }, SCAN_PERIOD);
    }

    public void stopScan() {
        if (!isScanning || scanner == null) return;

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            if (ActivityCompat.checkSelfPermission(context, Manifest.permission.BLUETOOTH_SCAN)
                    != PackageManager.PERMISSION_GRANTED) {
                return;
            }
        }

        scanner.stopScan(scanCallback);
        isScanning = false;
        Log.d(TAG, "⏹️ Stopped BLE scan");
    }

    private final ScanCallback scanCallback = new ScanCallback() {
        @Override
        public void onScanResult(int callbackType, ScanResult result) {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
                if (ActivityCompat.checkSelfPermission(context, Manifest.permission.BLUETOOTH_CONNECT)
                        != PackageManager.PERMISSION_GRANTED) {
                    return;
                }
            }

            BluetoothDevice device = result.getDevice();
            String deviceName = device.getName();

            if (deviceName != null && deviceName.toLowerCase().contains("cape")) {
                Log.d(TAG, "✅ Found ESP32 device: " + deviceName);
                stopScan();
                callback.onDeviceFound(device, deviceName);
            }
        }

        @Override
        public void onScanFailed(int errorCode) {
            isScanning = false;
            callback.onError("Scan failed with error: " + errorCode);
            Log.e(TAG, "❌ BLE scan failed: " + errorCode);
        }
    };

    public void connectToDevice(BluetoothDevice device) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            if (ActivityCompat.checkSelfPermission(context, Manifest.permission.BLUETOOTH_CONNECT)
                    != PackageManager.PERMISSION_GRANTED) {
                callback.onError("Bluetooth connect permission not granted");
                return;
            }
        }

        if (bluetoothGatt != null) {
            bluetoothGatt.close();
            bluetoothGatt = null;
        }

        Log.d(TAG, "🔗 Connecting to device: " + device.getAddress());
        bluetoothGatt = device.connectGatt(context, false, gattCallback);
    }

    private final BluetoothGattCallback gattCallback = new BluetoothGattCallback() {
        @Override
        public void onConnectionStateChange(BluetoothGatt gatt, int status, int newState) {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
                if (ActivityCompat.checkSelfPermission(context, Manifest.permission.BLUETOOTH_CONNECT)
                        != PackageManager.PERMISSION_GRANTED) {
                    return;
                }
            }

            if (newState == BluetoothProfile.STATE_CONNECTED) {
                isConnected = true;
                Log.d(TAG, "✅ Connected to GATT server");
                gatt.discoverServices();
            } else if (newState == BluetoothProfile.STATE_DISCONNECTED) {
                isConnected = false;
                Log.d(TAG, "❌ Disconnected from GATT server");
                handler.post(() -> callback.onDisconnected());
                cleanup();
            }
        }

        @Override
        public void onServicesDiscovered(BluetoothGatt gatt, int status) {
            if (status == BluetoothGatt.GATT_SUCCESS) {
                android.bluetooth.BluetoothGattService nusService = gatt.getService(NUS_SERVICE_UUID);
                if (nusService != null) {
                    txCharacteristic = nusService.getCharacteristic(NUS_TX_UUID);
                    rxCharacteristic = nusService.getCharacteristic(NUS_RX_UUID);

                    if (txCharacteristic != null && rxCharacteristic != null) {
                        Log.d(TAG, "🎯 NUS service characteristics found");
                        enableNotifications(gatt, txCharacteristic);
                        handler.post(() -> callback.onConnected());
                    } else {
                        handler.post(() -> callback.onError("NUS characteristics not found"));
                    }
                } else {
                    handler.post(() -> callback.onError("NUS service not found"));
                }
            } else {
                handler.post(() -> callback.onError("Service discovery failed: " + status));
            }
        }

        @Override
        public void onCharacteristicChanged(BluetoothGatt gatt, BluetoothGattCharacteristic characteristic) {
            if (characteristic.getUuid().equals(NUS_TX_UUID)) {
                byte[] data = characteristic.getValue();
                String message = new String(data, StandardCharsets.UTF_8);
                Log.d(TAG, "📥 Received: " + message);
                handler.post(() -> callback.onDataReceived(message));
            }
        }

        @Override
        public void onCharacteristicWrite(BluetoothGatt gatt, BluetoothGattCharacteristic characteristic, int status) {
            if (status == BluetoothGatt.GATT_SUCCESS) {
                Log.d(TAG, "📤 Write successful");
            } else {
                Log.e(TAG, "❌ Write failed: " + status);
            }
        }
    };

    private void enableNotifications(BluetoothGatt gatt, BluetoothGattCharacteristic characteristic) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            if (ActivityCompat.checkSelfPermission(context, Manifest.permission.BLUETOOTH_CONNECT)
                    != PackageManager.PERMISSION_GRANTED) {
                return;
            }
        }

        gatt.setCharacteristicNotification(characteristic, true);

        BluetoothGattDescriptor descriptor = characteristic.getDescriptor(CLIENT_CHARACTERISTIC_CONFIG);
        if (descriptor != null) {
            descriptor.setValue(BluetoothGattDescriptor.ENABLE_NOTIFICATION_VALUE);
            gatt.writeDescriptor(descriptor);
            Log.d(TAG, "📡 Notifications enabled for TX characteristic");
        }
    }

    public void sendCommand(String command) {
        if (!isConnected || bluetoothGatt == null || rxCharacteristic == null) {
            Log.w(TAG, "Cannot send - not connected");
            return;
        }

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            if (ActivityCompat.checkSelfPermission(context, Manifest.permission.BLUETOOTH_CONNECT)
                    != PackageManager.PERMISSION_GRANTED) {
                return;
            }
        }

        byte[] data = command.getBytes(StandardCharsets.UTF_8);
        rxCharacteristic.setValue(data);
        boolean success = bluetoothGatt.writeCharacteristic(rxCharacteristic);
        Log.d(TAG, "📤 Sending command: " + command + " - " + (success ? "queued" : "failed"));
    }

    public void disconnect() {
        if (bluetoothGatt != null) {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
                if (ActivityCompat.checkSelfPermission(context, Manifest.permission.BLUETOOTH_CONNECT)
                        != PackageManager.PERMISSION_GRANTED) {
                    return;
                }
            }
            bluetoothGatt.disconnect();
            bluetoothGatt.close();
            bluetoothGatt = null;
        }
        cleanup();
    }

    private void cleanup() {
        txCharacteristic = null;
        rxCharacteristic = null;
        isConnected = false;
    }

    public boolean isConnected() {
        return isConnected;
    }

    public void close() {
        stopScan();
        disconnect();
    }
}

File: app/src/main/java/com/example/movewithme/BleManager.kt
package com.example.movewithme

import android.bluetooth.BluetoothAdapter
import android.bluetooth.BluetoothDevice
import android.bluetooth.BluetoothGatt
import android.bluetooth.BluetoothGattCallback
import android.bluetooth.BluetoothGattCharacteristic
import android.bluetooth.BluetoothGattDescriptor
import android.bluetooth.BluetoothManager
import android.bluetooth.BluetoothProfile
import android.bluetooth.le.BluetoothLeScanner
import android.bluetooth.le.ScanCallback
import android.bluetooth.le.ScanFilter
import android.bluetooth.le.ScanResult
import android.bluetooth.le.ScanSettings
import android.content.Context
import android.os.Build
import android.os.Handler
import android.os.Looper
import android.os.ParcelUuid
import android.util.Log
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.channels.BufferOverflow
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asSharedFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import kotlinx.coroutines.suspendCancellableCoroutine
import kotlinx.coroutines.withTimeoutOrNull
import java.util.UUID
import kotlin.coroutines.resume

private const val TAG = "BleManager"

sealed class BleStatus {
    data object Disconnected : BleStatus()
    data object Scanning : BleStatus()
    data class Connecting(val device: BluetoothDevice) : BleStatus()
    data class Connected(val device: BluetoothDevice) : BleStatus()
    data class Error(val message: String) : BleStatus()
}

class BleManager(private val context: Context) {

    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Default)
    private val mainHandler = Handler(Looper.getMainLooper())
    private val bluetoothManager = context.getSystemService(BluetoothManager::class.java)
    private val bluetoothAdapter: BluetoothAdapter? = bluetoothManager?.adapter

    private val _connectionState = MutableStateFlow<BleStatus>(BleStatus.Disconnected)
    val connectionState: StateFlow<BleStatus> = _connectionState.asStateFlow()

    private val _notifications = MutableSharedFlow<String>(
        replay = 0,
        extraBufferCapacity = 64,
        onBufferOverflow = BufferOverflow.DROP_OLDEST
    )
    val notifications: Flow<String> = _notifications.asSharedFlow()

    private var bluetoothGatt: BluetoothGatt? = null
    private var rxCharacteristic: BluetoothGattCharacteristic? = null
    private var txCharacteristic: BluetoothGattCharacteristic? = null
    private var isScanning = false
    private var currentScanner: BluetoothLeScanner? = null
    private var currentScanCallback: ScanCallback? = null

    private val NUS_SERVICE_UUID: UUID = UUID.fromString("6E400001-B5A3-F393-E0A9-E50E24DCCA9E")
    private val NUS_RX_UUID: UUID = UUID.fromString("6E400002-B5A3-F393-E0A9-E50E24DCCA9E")
    private val NUS_TX_UUID: UUID = UUID.fromString("6E400003-B5A3-F393-E0A9-E50E24DCCA9E")
    private val CCCD_UUID: UUID = UUID.fromString("00002902-0000-1000-8000-00805F9B34FB")

    fun scanAndConnect() {
        scope.launch {
            val adapter = bluetoothAdapter
            if (adapter == null) {
                updateState(BleStatus.Error("Bluetooth not supported"))
                return@launch
            }
            if (!adapter.isEnabled) {
                updateState(BleStatus.Error("Bluetooth disabled"))
                return@launch
            }
            if (_connectionState.value is BleStatus.Connected || _connectionState.value is BleStatus.Connecting || _connectionState.value is BleStatus.Scanning) {
                return@launch
            }

            val device = withTimeoutOrNull(15000) { scanForDevice(adapter.bluetoothLeScanner) }
            if (device == null) {
                updateState(BleStatus.Error("Device not found"))
                return@launch
            }
            updateState(BleStatus.Connecting(device))
            connectToDevice(device)
        }
    }

    fun disconnect() {
        stopScan()
        mainHandler.post {
            bluetoothGatt?.let { gatt ->
                try {
                    gatt.disconnect()
                } catch (_: Throwable) {
                }
                try {
                    gatt.close()
                } catch (_: Throwable) {
                }
            }
            bluetoothGatt = null
            rxCharacteristic = null
            txCharacteristic = null
            updateState(BleStatus.Disconnected)
        }
    }

    fun sendCommand(cmd: String) {
        val gatt = bluetoothGatt
        val rx = rxCharacteristic
        if (gatt == null || rx == null) {
            Log.w(TAG, "Cannot send command, GATT not ready")
            return
        }
        val payload = (cmd + "\n").toByteArray(Charsets.UTF_8)
        mainHandler.post {
            rx.value = payload
            rx.writeType = BluetoothGattCharacteristic.WRITE_TYPE_DEFAULT
            val ok = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
                val status = gatt.writeCharacteristic(rx, payload, BluetoothGattCharacteristic.WRITE_TYPE_DEFAULT)
                status == android.bluetooth.BluetoothStatusCodes.SUCCESS
            } else {
                gatt.writeCharacteristic(rx)
            }
            Log.d(TAG, "sendCommand: '$cmd', writeCharacteristic returned $ok")
        }
    }

    fun close() {
        disconnect()
        scope.cancel()
    }

    private suspend fun scanForDevice(scanner: BluetoothLeScanner?): BluetoothDevice? {
        if (scanner == null) {
            updateState(BleStatus.Error("BLE scanner unavailable"))
            return null
        }
        updateState(BleStatus.Scanning)
        stopScan()
        return suspendCancellableCoroutine { continuation ->
            val filters = listOf(
                ScanFilter.Builder().setDeviceName("ESP32-GYRO-ON").build(),
                ScanFilter.Builder().setServiceUuid(ParcelUuid(NUS_SERVICE_UUID)).build()
            )
            val settings = ScanSettings.Builder()
                .setScanMode(ScanSettings.SCAN_MODE_LOW_LATENCY)
                .build()
            val callback = object : ScanCallback() {
                override fun onScanResult(callbackType: Int, result: ScanResult) {
                    val device = result.device
                    if (device != null && matchesTarget(device, result)) {
                        Log.d(TAG, "Found target device: ${device.address}")
                        stopAndResume(device)
                    }
                }

                override fun onScanFailed(errorCode: Int) {
                    Log.e(TAG, "Scan failed: $errorCode")
                    updateState(BleStatus.Error("Scan failed: $errorCode"))
                    stopAndResume(null)
                }

                override fun onBatchScanResults(results: MutableList<ScanResult>) {
                    results.firstOrNull { it.device != null && matchesTarget(it.device, it) }?.let {
                        onScanResult(ScanSettings.CALLBACK_TYPE_ALL_MATCHES, it)
                    }
                }

                private fun stopAndResume(device: BluetoothDevice?) {
                    if (!continuation.isActive) return
                    stopScanInternal(scanner, this)
                    continuation.resume(device)
                }
            }
            startScanInternal(scanner, filters, settings, callback)
            continuation.invokeOnCancellation {
                stopScanInternal(scanner, callback)
            }
        }
    }

    private fun matchesTarget(device: BluetoothDevice, result: ScanResult): Boolean {
        if (device.name == "ESP32-GYRO-ON") return true
        val uuids = result.scanRecord?.serviceUuids
        return uuids?.any { it.uuid == NUS_SERVICE_UUID } == true
    }

    private fun connectToDevice(device: BluetoothDevice) {
        mainHandler.post {
            stopScan()
            rxCharacteristic = null
            txCharacteristic = null
            bluetoothGatt?.close()
            bluetoothGatt = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                device.connectGatt(context, false, gattCallback, BluetoothDevice.TRANSPORT_LE)
            } else {
                device.connectGatt(context, false, gattCallback)
            }
        }
    }

    private fun startScanInternal(
        scanner: BluetoothLeScanner,
        filters: List<ScanFilter>,
        settings: ScanSettings,
        callback: ScanCallback
    ) {
        mainHandler.post {
            if (isScanning) return@post
            isScanning = true
            currentScanner = scanner
            currentScanCallback = callback
            scanner.startScan(filters, settings, callback)
        }
    }

    private fun stopScanInternal(scanner: BluetoothLeScanner, callback: ScanCallback) {
        mainHandler.post {
            if (!isScanning) return@post
            try {
                scanner.stopScan(callback)
            } catch (_: Throwable) {
            }
            isScanning = false
            currentScanner = null
            currentScanCallback = null
        }
    }

    private fun stopScan() {
        val scanner = currentScanner ?: bluetoothAdapter?.bluetoothLeScanner ?: return
        val callback = currentScanCallback ?: return
        stopScanInternal(scanner, callback)
    }

    private fun updateState(state: BleStatus) {
        _connectionState.value = state
    }

    private val gattCallback = object : BluetoothGattCallback() {
        override fun onConnectionStateChange(gatt: BluetoothGatt, status: Int, newState: Int) {
            super.onConnectionStateChange(gatt, status, newState)
            if (status != BluetoothGatt.GATT_SUCCESS) {
                Log.e(TAG, "Connection state change error: $status")
            }
            when (newState) {
                BluetoothProfile.STATE_CONNECTED -> {
                    Log.d(TAG, "Connected to GATT, discovering services")
                    updateState(BleStatus.Connecting(gatt.device))
                    gatt.discoverServices()
                }
                BluetoothProfile.STATE_DISCONNECTED -> {
                    Log.d(TAG, "Disconnected from GATT")
                    cleanupGatt()
                    if (status == BluetoothGatt.GATT_SUCCESS) {
                        updateState(BleStatus.Disconnected)
                    } else {
                        updateState(BleStatus.Error("Disconnected: $status"))
                    }
                }
            }
        }

        override fun onServicesDiscovered(gatt: BluetoothGatt, status: Int) {
            super.onServicesDiscovered(gatt, status)
            Log.d(TAG, "onServicesDiscovered: status=$status")
            if (status != BluetoothGatt.GATT_SUCCESS) {
                updateState(BleStatus.Error("Service discovery failed: $status"))
                return
            }

            Log.d(TAG, "Services found: ${gatt.services.map { it.uuid }}")
            val service = gatt.getService(NUS_SERVICE_UUID)
            val tx = service?.getCharacteristic(NUS_TX_UUID)
            val rx = service?.getCharacteristic(NUS_RX_UUID)

            if (service == null) {
                Log.e(TAG, "NUS service not found. Available services: ${gatt.services.map { it.uuid }}")
                updateState(BleStatus.Error("NUS service not found"))
                gatt.disconnect()
                return
            }

            if (tx == null || rx == null) {
                Log.e(TAG, "NUS characteristics not found. Service has: ${service.characteristics.map { it.uuid }}")
                updateState(BleStatus.Error("NUS characteristics not found"))
                gatt.disconnect()
                return
            }

            Log.d(TAG, "✓ Found NUS service and characteristics")
            bluetoothGatt = gatt
            rxCharacteristic = rx
            txCharacteristic = tx

            val notificationSet = gatt.setCharacteristicNotification(tx, true)
            Log.d(TAG, "setCharacteristicNotification for TX: $notificationSet")

            if (!notificationSet) {
                Log.e(TAG, "⚠ Failed to enable local notifications for TX characteristic")
            }

            val cccd = tx.getDescriptor(CCCD_UUID)
            if (cccd != null) {
                Log.d(TAG, "Writing ENABLE_NOTIFICATION_VALUE to CCCD descriptor...")
                cccd.value = BluetoothGattDescriptor.ENABLE_NOTIFICATION_VALUE
                val wrote = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
                    gatt.writeDescriptor(cccd, BluetoothGattDescriptor.ENABLE_NOTIFICATION_VALUE) == android.bluetooth.BluetoothStatusCodes.SUCCESS
                } else {
                    gatt.writeDescriptor(cccd)
                }
                Log.d(TAG, "CCCD write initiated: $wrote")
                if (!wrote) {
                    Log.e(TAG, "⚠ Failed to initiate CCCD descriptor write")
                }
            } else {
                Log.e(TAG, "⚠ CCCD descriptor not found on TX characteristic")
            }

            Log.d(TAG, "Requesting MTU increase to 512 bytes...")
            val mtuRequested = gatt.requestMtu(512)
            Log.d(TAG, "MTU request initiated: $mtuRequested")

            Log.d(TAG, "✓ BLE setup complete, updating state to Connected")
            updateState(BleStatus.Connected(gatt.device))
        }

        override fun onCharacteristicChanged(
            gatt: BluetoothGatt,
            characteristic: BluetoothGattCharacteristic
        ) {
            super.onCharacteristicChanged(gatt, characteristic)
            Log.d(TAG, "onCharacteristicChanged (deprecated): UUID=${characteristic.uuid}")
            if (characteristic.uuid != NUS_TX_UUID) {
                Log.w(TAG, "Ignoring notification from unexpected characteristic: ${characteristic.uuid}")
                return
            }
            val bytes = characteristic.value ?: run {
                Log.w(TAG, "Notification has null value")
                return
            }
            val text = try {
                bytes.toString(Charsets.UTF_8)
            } catch (e: Exception) {
                Log.e(TAG, "Failed to decode notification", e)
                return
            }
            Log.d(TAG, "✓ Notification received: '$text'")
            _notifications.tryEmit(text)
        }

        // Override for Android 13+ (API 33+)
        override fun onCharacteristicChanged(
            gatt: BluetoothGatt,
            characteristic: BluetoothGattCharacteristic,
            value: ByteArray
        ) {
            super.onCharacteristicChanged(gatt, characteristic, value)
            Log.d(TAG, "onCharacteristicChanged (new API): UUID=${characteristic.uuid}, bytes=${value.size}")
            if (characteristic.uuid != NUS_TX_UUID) {
                Log.w(TAG, "Ignoring notification from unexpected characteristic: ${characteristic.uuid}")
                return
            }
            val text = try {
                value.toString(Charsets.UTF_8)
            } catch (e: Exception) {
                Log.e(TAG, "Failed to decode notification", e)
                return
            }
            Log.d(TAG, "✓ Notification received: '$text'")
            _notifications.tryEmit(text)
        }

        override fun onMtuChanged(gatt: BluetoothGatt, mtu: Int, status: Int) {
            super.onMtuChanged(gatt, mtu, status)
            if (status == BluetoothGatt.GATT_SUCCESS) {
                Log.d(TAG, "✓✓✓ MTU changed SUCCESS! New MTU: $mtu bytes (payload: ${mtu - 3} bytes)")
            } else {
                Log.e(TAG, "⚠ MTU change failed: status=$status, using default MTU=23")
            }
        }

        override fun onDescriptorWrite(
            gatt: BluetoothGatt,
            descriptor: BluetoothGattDescriptor,
            status: Int
        ) {
            super.onDescriptorWrite(gatt, descriptor, status)
            if (status == BluetoothGatt.GATT_SUCCESS) {
                if (descriptor.uuid == CCCD_UUID && descriptor.characteristic.uuid == NUS_TX_UUID) {
                    Log.d(TAG, "✓✓✓ CCCD write SUCCESS! Notifications are now enabled. Waiting for data...")
                } else {
                    Log.d(TAG, "Descriptor write success: ${descriptor.uuid}")
                }
            } else {
                Log.e(TAG, "⚠⚠⚠ Descriptor write FAILED: status=$status, descriptor=${descriptor.uuid}")
            }
        }
    }

    private fun cleanupGatt() {
        rxCharacteristic = null
        txCharacteristic = null
        bluetoothGatt?.close()
        bluetoothGatt = null
    }
}